tests/integrations/resource_metering/test_read_keys.rs

```diff
@@ -0,0 +1,250 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use crate::resource_metering::test_suite::MockReceiverServer;
+
+use std::sync::atomic::AtomicU32;
+use std::sync::atomic::Ordering::SeqCst;
+use std::sync::Arc;
+use std::time::Duration;
+
+use crossbeam::channel::{unbounded, Receiver};
+use grpcio::{ChannelBuilder, Environment};
+use kvproto::coprocessor;
+use kvproto::kvrpcpb::*;
+use kvproto::resource_usage_agent::ResourceUsageRecord;
+use kvproto::tikvpb::*;
+use protobuf::Message;
+use resource_metering::Records;
+use test_coprocessor::{init_with_data, DAGSelect, ProductTable};
+use test_raftstore::*;
+use test_util::alloc_port;
+use tikv::coprocessor::Endpoint;
+use tikv::storage::Engine;
+use tikv_util::config::ReadableDuration;
+use tikv_util::worker::Builder as WorkerBuilder;
+use tikv_util::HandyRwLock;
+use tipb::SelectResponse;
+
+#[test]
+pub fn test_read_keys() {
+    // Create & start receiver server.
+    let (tx, rx) = unbounded();
+    let mut server = MockReceiverServer::new(tx);
+    let port = alloc_port();
+    let env = Arc::new(Environment::new(1));
+    server.start_server(port, env.clone());
+
+    // Create cluster.
+    let (_cluster, client, mut ctx) = new_cluster(port, env);
+
+    // Set resource group tag for enable resource metering.
+    ctx.set_resource_group_tag("TEST-TAG".into());
+
+    let mut ts = 0;
+
+    // Write 10 key-value pairs.
+    for n in 0..10 {
+        let n = n.to_string().into_bytes();
+        let (k, v) = (n.clone(), n);
+
+        // Prewrite.
+        ts += 1;
+        let prewrite_start_version = ts;
+        let mut mutation = Mutation::default();
+        mutation.set_op(Op::Put);
+        mutation.set_key(k.clone());
+        mutation.set_value(v.clone());
+        must_kv_prewrite(
+            &client,
+            ctx.clone(),
+            vec![mutation],
+            k.clone(),
+            prewrite_start_version,
+        );
+
+        // Commit.
+        ts += 1;
+        let commit_version = ts;
+        must_kv_commit(
+            &client,
+            ctx.clone(),
+            vec![k.clone()],
+            prewrite_start_version,
+            commit_version,
+            commit_version,
+        );
+    }
+
+    // PointGet
```

Comment:
The test stability may be not good, as it is purely possible that the scan is slow in some cases (consider it to be as slow as 200ms+ if there is a block cache miss LOL).. It could happen. You can fix it if unstable tests occurs later.

If you want to make it fully stable, a better way will be controlling ticks manually (clear -> scan -> take stats and clear -> ....). I'm fine it is not implemented now.

Reply:
At present, there is no good way to manually control the processing. This requires modification of the code structure, which can be proposed separately as an optimization.

Reply:
I just use wait/notify to avoid fixed thread sleep, as a simple but effective optimization.